### PR TITLE
allow variable substitution in all configuration strings

### PR DESCRIPTION
**Pull request by rbberger**
This allows using Spack configuration variables and environment variables in more places inside of the Spack configuration. This was originally started to allow `package_attributes` to make use of environment variables.

Example:
```yaml
# example: SITE_SPECIFIC_GIT_MIRROR=file:///some/local/shared/location
packages:
   foo:
      package_attributes:
         git: $SITE_SPECIFIC_GIT_MIRROR/foo.git
```

This was then further extended to perform variables substitution on all strings of the configuration. To maintain consistency with existing commands like `spack config get` and `spack config blame`, these were extended with a `--expand-vars` option which defaults to `False`. Internally all read configs are always expanded.

### CODE REVIEW

File: lib/spack/spack/repo.py [hunk 1]
```diff
@@ -684,8 +684,15 @@ def from_descriptors(
...
+
+        def replace_vars_in_package_attrs(attrs):
+            for a in ("git", "url"):
+                if a in attrs:
+                    attrs[a] = spack.util.path.substitute_path_variables(attrs[a])
+            return attrs
+
         overrides = {
-            pkg_name: data["package_attributes"]
+            pkg_name: replace_vars_in_package_attrs(data["package_attributes"])
```
Comment (psakievich) [inline lib/spack/spack/repo.py]: Is there a reason to NOT use `substitute_path_variables` directly?

File: lib/spack/spack/repo.py [hunk 2]
```diff
@@ -684,8 +684,15 @@ def from_descriptors(
     @staticmethod
     def from_config(config: spack.config.Configuration) -> "RepoPath":
         """Create a RepoPath from a configuration object."""
+
+        def replace_vars_in_package_attrs(attrs):
+            for a in ("git", "url"):
+                if a in attrs:
+                    attrs[a] = spack.util.path.substitute_path_variables(attrs[a])
+            return attrs
```
Comment (psakievich) [inline lib/spack/spack/repo.py]: I don't love the selective list here. I also think that if we are going to have a selective list then I think it should be an error to try and expand something not in the list.

It should also be added to the documentation so users know what the rules are.

**Review by psakievich — COMMENTED** [review]
I'm not sure how I feel about this change. I have cases where I would use this feature myself, but I also think it makes the package hash user environment dependent. That is at odds with many of the core design principles in Spack. 

At SNL we have typically handled mirror url issues via included configs i.e on network a, include a config with the url valid for network a, and likewise for network b.

Is there a shortcoming with that approach that this solves?

File: lib/spack/spack/repo.py (hunk 1 above)
Comment (rbberger) [reply lib/spack/spack/repo.py]: I've added the list for the package_attributes that I know would benefit from this. I'm selective because I didn't know if there are cases that would break otherwise.

**Comment by rbberger** [comment]
>  I have cases where I would use this feature myself, but I also think it makes the package hash user environment dependent. That is at odds with many of the core design principles in Spack.

Not sure I see where this is worse than having a conditional (e.g. via a `when` on
`$os` or `$env` variables) and include a file that manipulates the `package_attributes` accordingly. That would have the same "user environment" dependence.

For the cases I care about, I don't have a problem with the "user environment" dependence, because that part is taken care of by another tool. What this solves for me is that I can provide the path consistently once, and don't have to generate a YAML file with the necessary package_attributes and absolute paths anymore. It allows me to use the same generic configuration file, with different (dynamic) mirror locations.

**Comment by psakievich** [comment]
> Not sure I see where this is worse than having a conditional (e.g. via a when on
$os or $env variables) and include a file that manipulates the package_attributes accordingly. That would have the same "user environment" dependence.

Fair.  I suppose if the expansion is captured in the lock file then the environment is reproducible and transferable.

File: lib/spack/spack/repo.py (hunk 1 above)
Comment (psakievich) [reply lib/spack/spack/repo.py]: Let's brainstorm on if there are any that we'd want to squash this behavior on?  @becker33 and I noticed that you could set version :scream: via yaml, but the solver chokes because it is not the right object. So I think there is a level of exclusion that needs to be applied, but one level up at the accepting of attributes. 

I would prefer exclusion to inclusion, and error to pass for things that can't accept env variables so the users get a clear picture and don't have to chase down where an unexpanded  env variable string came from.

For attributes that can be strings, I can't really think of any off the top of my head that NEED to be excluded from this feature.

**Comment by becker33** [comment]
Currently the expansion would only be captured in the lock file by the package hash, but not in any reversible manner.

File: lib/spack/spack/repo.py (hunk 2 above)
Comment (becker33) [reply lib/spack/spack/repo.py]: We generally select which vars are path-type vars to apply `spack.util.path.substitute_path_variables`, we don't really have an automatic component to that.

I'm not sure why we couldn't just apply it to all values read from config -- there's nowhere that you need to use those variables without expansion that I can think of

Comment (psakievich) [reply lib/spack/spack/repo.py]: I like that idea. We have so many one off's where the configs break due to expansion. I think I average 1 per year for a config missing expansions.

Comment (psakievich) [reply lib/spack/spack/repo.py]: @rbberger would you be willing to replace this PR with @becker33 's idea? Would you be able to work this?  I can get it into my queue, but I don't want to leave you in limbo either.

Comment (rbberger) [reply lib/spack/spack/repo.py]: Apply it to everything? Sure.

Comment (rbberger) [reply lib/spack/spack/repo.py]: I've applied this now to all `package_attributes`. Or am I misunderstanding and you want it even more globally? In any case, I can work on it.

Comment (psakievich) [reply lib/spack/spack/repo.py]: More globally.  I was thinking something like here: https://github.com/spack/spack/blob/15e7b9edaeb516faab4512f95fa16d12eb412b90/lib/spack/spack/config.py#L714

or if there is something further upstream in the parsing process.

File: lib/spack/spack/util/path.py [hunk 3]
```diff
@@ -191,11 +191,17 @@ def repl(match):
...
+
 def substitute_path_variables(path):
     """Substitute config vars, expand environment vars, expand user home."""
     path = substitute_config_variables(path)
     path = os.path.expandvars(path)
-    path = os.path.expanduser(path)
+    # ignore cases where ~ and ~user aren't used with a path separator
+    # this is to prevent ~variant strings from being expanded
+    if HOME_PREFIX.match(path):
+        path = os.path.expanduser(path)
```
Comment (rbberger) [inline lib/spack/spack/util/path.py:204]: @becker33 here's one nasty corner case when doing substitution in all strings. Something like  `requires: ["~variant"]` gets expanded if `variant` matches a username. E.g. this only showed up with a `~shared` on Windows unit tests, probably because Python's `ntpath.expanduser` is implemented a bit different than `posixpath.expanduser`.

File: lib/spack/spack/repo.py (hunk 2 above)
Comment (rbberger) [reply lib/spack/spack/repo.py]: @psakievich @becker33 Implemented the substitution for the entire config and worked my way through the unit test failures. There were a few gotchas. Not sure if there is anything more I can do for the import check.

Comment (psakievich) [reply lib/spack/spack/repo.py]: Thanks @rbberger. Not sure how I missed this. I'll review tomorrow

File: lib/spack/spack/util/path.py (hunk 3 above)
Comment (becker33) [reply lib/spack/spack/util/path.py:204]: Sorry I didn't see this earlier. I was on extended travel from late july to late october.

I agree that is a very nasty corner case. I'm trying to look into whether the jsonschema can be the source of truth for whether config values have path resolutions applied -- that might be a nice middle-ground between the two PRs?

If not then this might still be the best approach -- I like the way it's ignoring this, we could probably also easily include `~` on its own as the entire string, since that can't ever match a variant obviously

**Comment by rbberger** [comment]
@psakievich @becker33 brought this one up to date again. Thoughts?

**Comment by rbberger** [comment]
Partially implemented in #52450. Closing for now.
